### PR TITLE
feat: add notification bot

### PR DESCRIPTION
**Pull request by seendsouza**

### CODE REVIEW

File: notification_discord_bot/constants.py [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+DB_PATH = os.getenv("DB_PATH", "db.json")
+LOGLEVEL = os.getenv("LOGLEVEL", "INFO").upper()
+F_LOGLEVEL = os.getenv("F_LOGLEVEL", "INFO").upper()
+S_LOGLEVEL = os.getenv("S_LOGLEVEL", "INFO").upper()
+
+
+# Static constants
+RENFT_BASE_URL = "https://v2.renft.io"
+SLEEP_TIME_S = 30
+DEFAULT_PAGE_SIZE = 100
```
Comment (nazariyv) [inline notification_discord_bot/constants.py:39]: I would set it to `1000`, Whoopi is approaching `3000` transactions. The graph supports page size of `1000` no problem.

File: notification_discord_bot/renft.py [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+    @abstractmethod
+    def has_been_observed(self) -> bool:
+        pass
+
+    @abstractmethod
+    def observe(self):
+        pass
+
+
+def get_lending_url(contract: ReNFTContract, lending_id: str):
```
Comment (nazariyv) [inline notification_discord_bot/renft.py]: not a biggie, since it works anyhow, but to be more accurate, we can set the type on `lending_id` to `Union[str, int]`

Comment (nazariyv) [reply notification_discord_bot/renft.py]: this is because a bunch of places `data.py` are supplying an `int` as a second argument into this function

File: notification_discord_bot/data.py [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+            inline=False,
+        )
+        return msg
+
+    def build_twitter_message(self):
+        nft = self.lending.nft()
+        rent_duration_unit = get_rent_duration_unit(self.contract)
+
+        if self.lending.daily_rent_price is not None:
+            msg = (
```
Comment (nazariyv) [inline notification_discord_bot/data.py]: this and the message on line `80` are essentially the same. We can combine them. For example, something like `f"{nft.name} lent for {'a fee of ' if self.lending.daily_rent_price is not None else ''}"`

File: notification_discord_bot/renft.py [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+    if contract.name == AVALANCHE_WHOOPI_CONTRACT_NAME:
+        return f"{RENFT_BASE_URL}/collections/castle-crush?lendingId={lending_id}"
+    contract_type = "collateral_free" if contract.is_collateral_free() else "collateral"
+    return f"{RENFT_BASE_URL}?ctx={contract_type}&lendingId={lending_id}"
+
+
+def get_profile_url(contract: "ReNFTContract", address: str):
```
Comment (nazariyv) [inline notification_discord_bot/renft.py]: seems like there is no circ dep issue here (since the function above and below are using it explicitly), so we can just drop the double quotes on `ReNFTContract`

File: notification_discord_bot/renft.py [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+    chain = contract.chain()
+    chain_mapping = {
```
Comment (nazariyv) [inline notification_discord_bot/renft.py]: totally up to you, but I would extract this mapping out into constants and call it `EXPLORER_CHAIN_MAPPING`, then we would avoid re-defining it on every call to this function and would have it neatly in consts in case any other code needs it in the future

File: notification_discord_bot/mixins.py [hunk 6]
```diff
@@ -0,0 +1,27 @@
+from abc import ABC, abstractmethod
+from typing import ClassVar
+
+from notification_discord_bot import constants
+
+
+class BlockchainContractMixin(ABC):
+    @property
+    @abstractmethod
+    def query_url(self) -> str:
```
Comment (nazariyv) [inline notification_discord_bot/mixins.py]: seems to me like `chain` can also be added here as `property` and `abstractmethod`. See `contracts.py` and how we are defining the `chain` function on each class, whereas we could be assigning the property just like `query_url`.

File: notification_discord_bot/contracts.py [hunk 7]
```diff
@@ -0,0 +1,265 @@
...
+        )
+        rentings = data.get("data", {}).get("rentings", [])
+        return [self.transform_renting(r) for r in rentings]
+
+    def is_collateral_free(self) -> bool:
+        return True
+
+
+class EthereumAzraelContract(mixins.EthereumAzraelMixin, AzraelContract):
+    name = ETHEREUM_AZRAEL_CONTRACT_NAME
```
Comment (nazariyv) [inline notification_discord_bot/contracts.py:258]: It's a little hard to follow with the mixins. What advantages does this inheritance have over simply inheriting `mixins.EthereumAzraelMixin` directly in the `AzraelContract` class in place of the `BlockChainContract` mixin? We would effectively remove all four classes here. I have searched through the whole of the repository, and the only places where we use `BlockChainContract` mixin is precisely in the definitions of the classes above.

**Review by nazariyv — CHANGES_REQUESTED** [review]

File: notification_discord_bot/constants.py (hunk 1 above)
Comment (seendsouza) [reply notification_discord_bot/constants.py:39]: Discussed https://discord.com/channels/842471156413235261/972150826103627846/1020069741529927690

File: notification_discord_bot/renft.py (hunk 5 above)
Comment (seendsouza) [reply notification_discord_bot/renft.py]: Addressed, but kept in renft because it requires Chain as key

File: notification_discord_bot/mixins.py (hunk 6 above)
Comment (seendsouza) [reply notification_discord_bot/mixins.py]: Addressed

File: notification_discord_bot/contracts.py (hunk 7 above)
Comment (seendsouza) [reply notification_discord_bot/contracts.py:258]: Addressed

File: notification_discord_bot/renft.py (hunk 4 above)
Comment (seendsouza) [reply notification_discord_bot/renft.py]: Addressed

File: notification_discord_bot/data.py (hunk 3 above)
Comment (seendsouza) [reply notification_discord_bot/data.py]: Addressed

File: notification_discord_bot/renft.py (hunk 2 above)
Comment (seendsouza) [reply notification_discord_bot/renft.py]: Addressed

**Comment by nazariyv** [comment]
We are getting:

```
2780 |:| INFO |:| 2022-09-25 14:42:29.364 |:| Discord is enabled: True
2780 |:| INFO |:| 2022-09-25 14:42:29.365 |:| Twitter is enabled: True
Traceback (most recent call last):
  File "/home/shredder/git/notification-discord-bot/notification_discord_bot/main.py", line 73, in <module>
    main()
  File "/home/shredder/git/notification-discord-bot/notification_discord_bot/main.py", line 67, in main
    check_for_updates(msg_sender)
  File "/home/shredder/git/notification-discord-bot/notification_discord_bot/main.py", line 55, in check_for_updates
    msg_sender.send_discord_message(discord_message)
  File "/home/shredder/git/notification-discord-bot/notification_discord_bot/main.py", line 36, in send_discord_message
    self.discord_webhook.send(embed=msg)
  File "/home/shredder/.cache/pypoetry/virtualenvs/notification-discord-bot-eXHKOFjM-py3.10/lib/python3.10/site-packages/discord/webhook.py", line 989, in send
    return self._adapter.execute_webhook(wait=wait, file=file, files=files, payload=payload)
  File "/home/shredder/.cache/pypoetry/virtualenvs/notification-discord-bot-eXHKOFjM-py3.10/lib/python3.10/site-packages/discord/webhook.py", line 164, in execute_webhook
    maybe_coro = self.request('POST', url, multipart=multipart, payload=data, files=files_to_pass)
  File "/home/shredder/.cache/pypoetry/virtualenvs/notification-discord-bot-eXHKOFjM-py3.10/lib/python3.10/site-packages/discord/webhook.py", line 363, in request
    raise HTTPException(r, response)
discord.errors.HTTPException: 400 Bad Request (error code: 50035): Invalid Form Body
In embeds.0.thumbnail.url: Scheme "ipfs" is not supported. Scheme must be one of ('http', 'https').
```

when trying to run this. Must be because it is trying to post ipfs media on twitter.

**Comment by seendsouza** [comment]
Above comment addressed
